text/0000-process-stdio-redirection.md

```diff
@@ -0,0 +1,106 @@
+- Feature Name: process-stdio-redirection
+- Start Date: 2015-04-10
+- RFC PR:
+- Rust Issue:
+
+# Summary
+
+Update the standard library with a cross-platform high level API for redirecting
+stdio of child processes to any opened file handle or equivalent.
+
+# Motivation
+
+The current API in `std::process` allows the usage of raw file descriptors or
+HANDLEs for child process stdio redirection by leveraging the `FromRaw{Fd,
+Handle}` and `AsRaw{Fd, Handle}` trait implementations on `Stdio` and
+`ChildStd{In, Out, Err}`, respectively.
+
+Unfortunately, since the actual methods pertaining to `FromRaw*` and `AsRaw*`
+are OS specific, their usage requires either constant `cfg` checks or OS
+specific code duplication. Moreover, the conversion to `Stdio` is `unsafe` and
+requires the caller to ensure the OS handle remains valid until spawning the
```

Comment:
The API is actually a little subtle in this regard, but the contract of `FromRawFd` is that it takes ownership of the file descriptor (or handle) in question, so the unsafety is based on ownership, not on the lifetime of the resource itself. Specifically, the various primitives in the standard library provide the guarantee that they are the sole owner of the fd/handle in question, and this being safe would otherwise be a violation of that.

Also, if we assume that a `IntoRawFd` trait exists, using the raw underlying interfaces may not be so bad as you'd just have two functions (unix/windows) doing `Stdio::from_raw_fd(file.into_raw_fd())`. I suspect that adding traits like `IntoRawFd` are somewhat inevitable.